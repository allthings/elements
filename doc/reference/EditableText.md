<!-- 
This is an auto-generated markdown. 
You can change it in "src/EditableText/EditableText.tsx" and run build:docs to update this file.
-->
# EditableText

## Usage
| Name        | Type           | Description  |
| ----------- |:--------------:| ------------:|
|children|string, (string & {}), (string & ReactElement<any, string, ((props: any) => ReactElement<any, string, ..., (new (props: any) => Component<any, any, any>)>, null), (new (props: any) => Component<any, any, any>)>), (string & ReactNodeArray), (string & ReactPortal)|Text content passed to the component as children
|decorationColor|string|Color of the emulated text decoration (border bottom)
|icon|"alarm", "alarm-filled", "armchair", "armchair-filled", "arrow-down", "arrow-down-filled", "arrow-left", "arrow-left-filled", "arrow-right", "arrow-right-filled", ... 147 more ...|Icon displayed on the right side of the text
|iconColor|string|Color of the icon
|htmlElement|string|
|alignH|"none", "start", "center", "end", "space-around", "space-between"|horizontal alignment
|alignV|"none", "start", "center", "end", "stretch"|vertical alignment
|direction|"row", "column", "row-reverse", "column-reverse"|direction
|fill|boolean|Passing true, will make the view fill out available space
|wrap|"wrap", "inherit", "initial", "nowrap", "wrap-reverse"|Defining how children will wrap
|flex|number, "flex", "none", "initial", "nogrow", "grow", "auto", "noshrink"|Flex values, can be 5, 10, 15 ... 100 or 33, 66
