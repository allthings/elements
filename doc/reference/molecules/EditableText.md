<!-- 
This is an auto-generated markdown. 
You can change it in "src/molecules/EditableText.tsx" and run build:docs to update this file.
-->
# EditableText

## Usage
| Name        | Type           | Description  |
| ----------- |:--------------:| ------------:|
|children|string, (string & {}), (string & ReactElement<any, string, ((props: any) => ReactElement<any, string, ..., (new (props: any) => Component<any, any, any>)>, null), (new (props: any) => Component<any, any, any>)>), (string & ReactNodeArray), (string & ReactPortal)|Text content passed to the component as children
|decorationColor|string|Color of the emulated text decoration (border bottom)
|icon|"alarm", "alarm-filled", "armchair", "armchair-filled", "arrow-down", "arrow-down-filled", "arrow-left", "arrow-left-filled", "arrow-right", "arrow-right-filled", ... 147 more ...|Icon displayed on the right side of the text
|iconColor|string|Color of the icon
