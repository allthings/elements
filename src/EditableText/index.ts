export { default } from './EditableText'
export * from './EditableText'
