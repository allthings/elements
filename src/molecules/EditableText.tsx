import { ColorPalette } from '@allthings/colors'
import { css } from 'glamor'
import React, { KeyboardEvent } from 'react'

import Icon, { IconType } from '../atoms/Icon'
import Text from '../atoms/Text'
import View from '../atoms/View'

const DEFAULT_DECORATION_COLOR = 'grey'
const DEFAULT_ICON = 'edit'
const DEFAULT_ICON_COLOR = 'black'
const ENTER = 'Enter'

const STYLES = {
  editable: (decorationColor: string) =>
    css({
      borderBottom: `1px dashed ${ColorPalette[decorationColor]}`,
      ':focus': {
        borderBottom: `1px solid ${ColorPalette[decorationColor]}`,
      },
    }),
  icon: css({
    marginLeft: '4px',
  }),
}

const EditableText = ({
  children,
  decorationColor = DEFAULT_DECORATION_COLOR,
  icon = DEFAULT_ICON,
  iconColor = DEFAULT_ICON_COLOR,
  ...props
}: IEditableTextProps) => (
  <View alignV="center" direction="row">
    <Text
      block
      contentEditable
      // Disable multi-line editing.
      onKeyPress={(event: KeyboardEvent<HTMLElement>) =>
        event.key === ENTER && event.preventDefault()
      }
      spellCheck={false}
      // https://reactjs.org/docs/dom-elements.html#suppresscontenteditablewarning
      suppressContentEditableWarning
      {...STYLES.editable(decorationColor)}
      {...props}
    >
      {children}
    </Text>
    <View>
      <Icon
        color={ColorPalette[iconColor]}
        name={icon}
        size={12}
        {...STYLES.icon}
      />
    </View>
  </View>
)

interface IEditableTextProps {
  /** Text content passed to the component as children */
  children?: string
  /** Color of the emulated text decoration (border bottom) */
  decorationColor?: string
  /** Icon displayed on the right side of the text */
  icon?: IconType
  /** Color of the icon */
  iconColor?: string
}

export default EditableText
