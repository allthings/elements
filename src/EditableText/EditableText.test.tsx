import * as React from 'react'
import ReactDOM from 'react-dom'
import EditableText from './EditableText'
import { ColorPalette } from '@allthings/colors'
import { shallow, mount } from 'enzyme'
import { fromDOMNode } from 'jest-glamor-react'

const EDITABLE = 'div[contentEditable]'
const ICON = 'view'
const ICON_COLOR = 'red'
const ICON_COLOR_FROM_PALETTE = ColorPalette.red
const TEXT = 'Edit me!'

describe('Test the editable text component', () => {
  it('should be editable', () => {
    const wrapper = mount(<EditableText>{TEXT}</EditableText>)
    expect(wrapper).toMatchSnapshot()
    // Enzyme doesn't provide a way to modify contentEditable!
    // https://github.com/airbnb/enzyme/issues/927
    wrapper.setProps({ children: TEXT + '?' })
    expect(wrapper.find(EDITABLE).text()).toBe(TEXT + '?')
    wrapper.find(EDITABLE).simulate('keyDown', { key: 'Enter' })
    expect(wrapper.find(EDITABLE).text()).toBe(TEXT + '?')
  })
  it('should be using a custom icon', () => {
    const wrapper = shallow(<EditableText icon={ICON}>{TEXT}</EditableText>)
    expect(wrapper).toMatchSnapshot()
    expect(wrapper.find('Icon').prop('name')).toBe(ICON)
  })
  it('should be using a custom icon color', () => {
    const wrapper = shallow(
      <EditableText icon={ICON} iconColor={ICON_COLOR}>
        {TEXT}
      </EditableText>,
    )
    expect(wrapper).toMatchSnapshot()
    expect(wrapper.find('Icon').prop('color')).toBe(ICON_COLOR_FROM_PALETTE)
  })
  it('should be using a custom decoration color', () => {
    // Ideally this should be working:
    // https://github.com/kentcdodds/jest-glamor-react#tohavestyleruleproperty-value
    // But it never get any CSS rule... Instead, let's snapshot-diff!
    // https://github.com/kentcdodds/jest-glamor-react#integration-with-snapshot-diff
    const renderer = (ui: React.ReactElement) => {
      const div = document.createElement('div')
      ReactDOM.render(ui, div)
      return div.children[0]
    }
    const wrapper = renderer(<EditableText>{TEXT}</EditableText>)
    const wrapperWithDecoration = renderer(
      <EditableText decorationColor={ICON_COLOR}>{TEXT}</EditableText>,
    )
    // workaround for probably a linter bug
    // if we inlined the expect expression the linter would fail with
    // > TypeError: Cannot read property 'name' of undefined
    const exp = expect(fromDOMNode(wrapper)) as any
    exp.toMatchDiffSnapshot(fromDOMNode(wrapperWithDecoration))
  })
})
