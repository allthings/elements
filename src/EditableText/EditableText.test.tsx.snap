// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test the editable text component should be editable 1`] = `
.css-0,
[data-css-0] {
  margin-left: 4px;
}

.css-1,
[data-css-1] {
  border-bottom: 1px dashed #95a5a5;
}

.css-1:focus,
[data-css-1]:focus {
  border-bottom: 1px solid #95a5a5;
}

.css-3,
[data-css-3] {
  box-sizing: border-box;
  align-content: center;
  align-items: center;
  display: -webkit-box;
  display: -moz-box;
  display: -ms-flexbox;
  display: -webkit-flex;
  display: flex;
  flex-direction: row;
  justify-content: flex-start;
  flex: 0 0 auto;
  -webkit-align-content: center;
  -webkit-box-align: center;
  -webkit-align-items: center;
  -webkit-box-orient: horizontal;
  -webkit-box-direction: normal;
  -webkit-flex-direction: row;
  -webkit-box-pack: start;
  -webkit-justify-content: flex-start;
  -webkit-flex: 0 0 auto;
}

.css-4,
[data-css-4] {
  display: block;
  font-family: "Open Sans", Helvetica, Arial, sans-serif;
  font-weight: 400;
  font-size: 14px;
  -webkit-font-smoothing: antialiased;
  -moz-osx-font-smoothing: grayscale;
  color: #333333;
}

.css-5,
[data-css-5] {
  box-sizing: border-box;
  flex: 0 0 auto;
  -webkit-flex: 0 0 auto;
}

.css-6,
[data-css-6] {
  width: 12px;
  height: 12px;
  stroke: #000000;
}

<ForwardRef
  alignV="center"
  direction="row"
>
  <div
    data-css-3=""
  >
    <ForwardRef
      contentEditable={true}
      data-css-1=""
      data-css-4=""
      onKeyPress={[Function]}
      spellCheck={false}
      suppressContentEditableWarning={true}
    >
      <div
        contentEditable={true}
        data-css-1=""
        data-css-5=""
        data-css-4=""
        onKeyPress={[Function]}
        spellCheck={false}
        suppressContentEditableWarning={true}
      >
        Edit me!
      </div>
    </ForwardRef>
    <ForwardRef>
      <div
        data-css-5=""
      >
        <ForwardRef
          alignH="center"
          alignV="center"
          dangerouslySetInnerHTML={
            Object {
              "__html": "",
            }
          }
          data-css-6=""
          data-css-0=""
        >
          <div
            dangerouslySetInnerHTML={
              Object {
                "__html": "",
              }
            }
            data-css-6=""
            data-css-0=""
            data-css-5=""
          />
        </ForwardRef>
      </div>
    </ForwardRef>
  </div>
</ForwardRef>
`;

exports[`Test the editable text component should be using a custom decoration color 1`] = `
Snapshot Diff:
- First value
+ Second value

@@ -3,20 +3,10 @@
    margin-left: 4px;
  }

  .css-1,
  [data-css-1] {
-   border-bottom: 1px dashed #95a5a5;
- }
-
- .css-1:focus,
- [data-css-1]:focus {
-   border-bottom: 1px solid #95a5a5;
- }
-
- .css-3,
- [data-css-3] {
    box-sizing: border-box;
    align-content: center;
    align-items: center;
    display: -webkit-box;
    display: -moz-box;
@@ -35,52 +25,62 @@
    -webkit-box-pack: start;
    -webkit-justify-content: flex-start;
    -webkit-flex: 0 0 auto;
  }

- .css-4,
- [data-css-4] {
+ .css-2,
+ [data-css-2] {
    display: block;
    font-family: "Open Sans", Helvetica, Arial, sans-serif;
    font-weight: 400;
    font-size: 14px;
    -webkit-font-smoothing: antialiased;
    -moz-osx-font-smoothing: grayscale;
    color: #333333;
  }

- .css-5,
- [data-css-5] {
+ .css-3,
+ [data-css-3] {
    box-sizing: border-box;
    flex: 0 0 auto;
    -webkit-flex: 0 0 auto;
  }

- .css-6,
- [data-css-6] {
+ .css-4,
+ [data-css-4] {
    width: 12px;
    height: 12px;
    stroke: #000000;
+ }
+
+ .css-5,
+ [data-css-5] {
+   border-bottom: 1px dashed #e84c3d;
+ }
+
+ .css-5:focus,
+ [data-css-5]:focus {
+   border-bottom: 1px solid #e84c3d;
  }

  <div
-   data-css-3=""
+   data-css-1=""
  >
    <div
      contenteditable="true"
-     data-css-1=""
      data-css-5=""
-     data-css-4=""
+     data-css-3=""
+     data-css-2=""
      spellcheck="false"
    >
      Edit me!
    </div>
    <div
-     data-css-5=""
+     data-css-3=""
    >
      <div
-       data-css-6=""
+       data-css-4=""
        data-css-0=""
-       data-css-5=""
+       data-css-3=""
      />
    </div>
  </div>
`;

exports[`Test the editable text component should be using a custom icon 1`] = `
.css-0,
[data-css-0] {
  margin-left: 4px;
}

.css-1,
[data-css-1] {
  border-bottom: 1px dashed #95a5a5;
}

.css-1:focus,
[data-css-1]:focus {
  border-bottom: 1px solid #95a5a5;
}

<ForwardRef
  alignV="center"
  direction="row"
>
  <Text
    block={true}
    contentEditable={true}
    data-css-1=""
    onKeyPress={[Function]}
    spellCheck={false}
    suppressContentEditableWarning={true}
  >
    Edit me!
  </Text>
  <ForwardRef>
    <Icon
      color="#000000"
      data-css-0=""
      name="view"
      size={12}
    />
  </ForwardRef>
</ForwardRef>
`;

exports[`Test the editable text component should be using a custom icon color 1`] = `
.css-0,
[data-css-0] {
  margin-left: 4px;
}

.css-1,
[data-css-1] {
  border-bottom: 1px dashed #95a5a5;
}

.css-1:focus,
[data-css-1]:focus {
  border-bottom: 1px solid #95a5a5;
}

<ForwardRef
  alignV="center"
  direction="row"
>
  <Text
    block={true}
    contentEditable={true}
    data-css-1=""
    onKeyPress={[Function]}
    spellCheck={false}
    suppressContentEditableWarning={true}
  >
    Edit me!
  </Text>
  <ForwardRef>
    <Icon
      color="#e84c3d"
      data-css-0=""
      name="view"
      size={12}
    />
  </ForwardRef>
</ForwardRef>
`;
